src/app/providers/LocalSettingsProvider.tsx

```diff
@@ -0,0 +1,46 @@
+import { FC, PropsWithChildren, useEffect, useState } from 'react'
+import { LocalSettingsContext, LocalSettingsProviderContext } from './LocalSettingsContext'
+import { storage } from '../utils/storage'
+import { StorageKeys } from '../../types/storage'
+import { TableAgeType } from '../../types/table-age-type'
+import { defaultLocalSettings, LocalSettings } from '../../types/local-settings'
+
+const localStorage = storage()
+
+type LocalSettingsState = LocalSettings & { dirty: boolean }
+
+const localSettingsInitialState: LocalSettingsState = {
+  ...defaultLocalSettings,
+  ...(localStorage.get(StorageKeys.LocalSettings) ?? {}),
+  dirty: false,
+}
+
+export const LocalSettingsContextProvider: FC<PropsWithChildren> = ({ children }) => {
+  const [state, setState] = useState<LocalSettingsState>(localSettingsInitialState)
+
+  useEffect(() => {
+    const { dirty, ...settings } = state
+    if (dirty) {
+      localStorage.set(StorageKeys.LocalSettings, settings)
+      setState(prevState => ({
+        ...prevState,
+        dirty: false,
+      }))
+    }
+  }, [state])
+
+  const setAgeHeaderType = (ageHeaderType: TableAgeType) => {
+    setState(prevState => ({
+      ...prevState,
+      ageHeaderType,
+      dirty: true,
+    }))
+  }
```

Comment:
I would refactor this as well to `setLocalSetting(key: keyof LocalSettings, value)`, not sure if type of value can be inferred without generic type here. And probably remove the dirty flag(and above `useEffect`), as it calls `setState` twice, I would prefer if we would avoid unnecessary re-renders.

Reply:
Updated accordingly. (No more `dirty` and `useEffect()`, also there is a `setLocalSetting()` function to be called by the individual wrappers, which we still need for type safefy.

Reply:
With the current setup, care must be taken when adding new wrapper functions to use the correct types (matching with those defined in the type `LocalSetting`.)

Reply:
It's all fine now, no more type safety issues. (Using `Partial<LocalSettings>` instead of `keyof LocalSettings`)

Reply:
I meant something like this:
```typescript
  const setValues = <T extends keyof LocalSettings>(key: T, value: LocalSettings[T]) => {
    setState(prevState => ({
      ...prevState,
      ...{ [key]: value },
    }))

    localStorage.set(StorageKeys.LocalSettings, {
      ...state,
      ...{ [key]: value },
    })
  }

  const setAgeHeaderType = (ageHeaderType: TableAgeType) => setValues('ageHeaderType', ageHeaderType)
```
And just expose `setValue`(probably rename it), don't see a reason why we would need to define "wrapper" function for every key:value pair we add to LocalSettings map.

Reply:
Yeah, refactored accordingly.